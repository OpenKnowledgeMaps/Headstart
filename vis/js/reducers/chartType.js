export default chartType = (state = null, action) => {
    switch (action.type) {
      case "SET_NORMAL":
        return 'normal';
      case "SET_STREAMGRAPH":
          return 'streamgraph'
      default:
        return state;
    }
  };
