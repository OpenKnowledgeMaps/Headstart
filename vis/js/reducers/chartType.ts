export const STREAMGRAPH_MODE = "streamgraph";
export const KNOWLEDGEMAP_MODE = "knowledgeMap";

type ChartType = typeof STREAMGRAPH_MODE | typeof KNOWLEDGEMAP_MODE;

const chartType = (state: ChartType = KNOWLEDGEMAP_MODE, action: any) => {
  if (action.canceled) {
    return state;
  }
  
  switch (action.type) {
    case "INITIALIZE":
      return action.configObject.is_streamgraph
        ? STREAMGRAPH_MODE
        : KNOWLEDGEMAP_MODE;
    default:
      return state;
  }
};

export default chartType;
