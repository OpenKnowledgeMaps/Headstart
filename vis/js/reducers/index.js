/**
 * The main reducer where all subreducers are combined into one.
 */
import { combineReducers } from "redux";
import zoom from "./zoom";

export default combineReducers({
  zoom,
});
