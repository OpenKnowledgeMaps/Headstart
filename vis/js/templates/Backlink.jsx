import React from "react";

const Backlink = ({ label, onClick }) => {
  return (
    // html template starts here
    <p id="backlink" className="backlink">
      <a className="underline" onClick={onClick}>
        {label}
      </a>
    </p>
    // html template ends here
  );
};

export default Backlink;
