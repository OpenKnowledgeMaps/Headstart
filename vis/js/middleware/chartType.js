import { STREAMGRAPH_MODE } from "../reducers/chartType";

/**
 * Creates a middleware that adds a boolean 'isStreamgraph' to each action.
 */
const chartTypeMiddleware = ({ getState }) => {
  return (next) => (action) => {
    action.isStreamgraph = getState().chartType === STREAMGRAPH_MODE;
    return next(action);
  };
};

export default chartTypeMiddleware;
