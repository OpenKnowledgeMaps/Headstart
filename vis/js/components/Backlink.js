import React from "react";
import { connect } from "react-redux";
import { zoomOut } from "../actions";
import { STREAMGRAPH_MODE } from "../reducers/chartType";

import BacklinkTemplate from "../templates/Backlink";

import { createAnimationCallback } from "../utils/eventhandlers";

export const Backlink = ({
  hidden = false,
  streamgraph = false,
  onClick,
  localization = {},
}) => {
  if (hidden) {
    return null;
  }

  const handleOnClick = () => {
    if (onClick && typeof onClick === "function") {
      onClick();
    }
  };

  return (
    <BacklinkTemplate
      label={localization.backlink}
      onClick={handleOnClick}
      className={streamgraph ? "backlink backlink-streamgraph" : "backlink"}
    />
  );
};

const mapStateToProps = (state) => ({
  localization: state.localization,
  hidden: !state.zoom,
  streamgraph: state.chartType === STREAMGRAPH_MODE,
});

const mapDispatchToProps = (dispatch) => ({
  onClick: () => dispatch(zoomOut(createAnimationCallback(dispatch))),
});

export default connect(mapStateToProps, mapDispatchToProps)(Backlink);
