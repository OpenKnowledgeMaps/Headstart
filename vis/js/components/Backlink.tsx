
import React from "react";
import { connect } from "react-redux";
import { zoomOut } from "../actions";
import { STREAMGRAPH_MODE } from "../reducers/chartType";

import BackLinkTemplate from "../templates/Backlink";

import { createAnimationCallback } from "../utils/eventhandlers";
import useMatomo from "../utils/useMatomo";
import { Dispatch } from "redux";

export interface BackLinkProps {
  hidden?: boolean;
  streamgraph?: boolean;
  onClick: () => void;
  localization: any
}

export const BackLink = ({
  hidden = false,
  streamgraph = false,
  onClick,
  localization = {},
}: BackLinkProps) => {
  const { trackEvent } = useMatomo();

  if (hidden) {
    return null;
  }

  const handleOnClick = () => {
    trackEvent("Title & Context line", "Zoom out", "Backlink");
    if (onClick && typeof onClick === "function") {
      onClick();
    }
  };

  return (
    <BackLinkTemplate
      label={localization.backlink}
      onClick={handleOnClick}
      className={streamgraph ? "backlink backlink-streamgraph" : "backlink"}
    />
  );
};

const mapStateToProps = (state: any) => ({
  localization: state.localization,
  hidden: !state.zoom,
  streamgraph: state.chartType === STREAMGRAPH_MODE,
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  onClick: () => dispatch(zoomOut(createAnimationCallback(dispatch))),
});

export default connect(mapStateToProps, mapDispatchToProps)(BackLink);
