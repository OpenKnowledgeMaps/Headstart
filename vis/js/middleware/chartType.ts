// @ts-nocheck

import { STREAMGRAPH_MODE } from "../reducers/chartType";

/**
 * Creates a middleware that adds a boolean 'isStreamgraph' to each action.
 */
const chartTypeMiddleware = ({ getState }) => {
  return (next) => (action: any) => {
    action.isStreamgraph = getState().chartType === STREAMGRAPH_MODE;
    return next(action);
  };
};

export default chartTypeMiddleware;
