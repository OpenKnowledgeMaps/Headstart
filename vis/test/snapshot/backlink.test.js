import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";

import Backlink from "../../js/components/Backlink";

import { STREAMGRAPH_MODE } from "../../js/reducers/chartType";

const mockStore = configureStore([]);

describe("Backlink component snapshot", () => {
  it("matches as knowledgeMap snapshot", () => {
    const store = mockStore({
      zoom: true,
      chartType: "knowledgeMap",
      localization: {
        backlink: "← Back to overview",
      },
    });
    const tree = renderer
      .create(
        <Provider store={store}>
          <Backlink />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("matches as streamgraph snapshot", () => {
    const store = mockStore({
      zoom: true,
      chartType: STREAMGRAPH_MODE,
      localization: {
        backlink: "← Back to overview",
      },
    });
    const tree = renderer
      .create(
        <Provider store={store}>
          <Backlink />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
