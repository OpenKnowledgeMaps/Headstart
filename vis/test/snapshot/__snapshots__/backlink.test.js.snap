// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backlink component snapshot matches as snapshot 1`] = `
<p
  className="backlink"
  id="backlink"
>
  <a
    className="underline"
    onClick={[Function]}
  >
    ‚Üê Back to overview
  </a>
</p>
`;
